always-hydrate.md

```diff
@@ -0,0 +1,133 @@
+
+Today, a user can partially communicate this desire with the `gvfs prefetch --files` or `gvfs
+prefetch --folders` commands to bulk download the files they care about. And they can further specify
+the `--hydrate` flag to either command to bulk hydrate those files into the working directory.
+
+However, this is not a great experience because it's just a one-time prefetch and hydrate. What that
+means is that the next time the projection changes, e.g. after a `git checkout`, anywhere from some
+to all of those files can become dehydrated, and the user will again pay the cost of file-by-file
+downloads while trying to build.
+
+With this feature, we want the developer to make one gesture and communicate that a set of files
+should always be kept hydrated, even after the projection changes. That being said, a secondary goal
+is to not add any significant time to commands like `git checkout` that change the projection, so the
+implementation should ensure that these are done as background downloads that do not block these
+commands.
+
+## UX
+This is really an extension of the existing prefetch functionality. As such, it makes sense to tie
+the UX to the existing `gvfs prefetch` verb.
+
+Side note on the existing prefetch verb: it's really two different commands in one, and it's
+unfortunate that they are combined. The command `gvfs prefetch --commits` is entirely different in
+its goals, usage, and implementation than either of `gvfs prefetch --files` or `gvfs prefetch 
+--folders` (and the latter two can be combined with each other, but not with `--commits`). Ideally we
+would have had `gvfs prefetch-commits` and `gvfs prefetch-files` as two separate commands, which
+would have then led to a cleaner design for this new addition. I'm open to other suggestions here
+if anyone sees a better pattern for extending the current verbs.
+
+### Proposed UX for always-hydrate commands:
+Adding new entries: 
+
+Add an `--always-hydrate` flag to the existing `--files` and `--folders` variants of `gvfs prefetch`.
+This will communicate that whatever list of folders, files, and wildcards that were passed in to the
+command should be saved and reapplied after every projection change
+
+Syntax:
+
+```
+    gvfs prefetch 
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      [--folders-list <path of file containing newline-delimited paths of folders, no wildcards>] 
+      [--[always-]hydrate]
+```
+
+The only change from today's syntax is the optional `[always-]` on the hydrate flag. 
+`--always-hydrate` implies `--hydrate`.
+
+Passing in the `--always-hydrate` flag has all the same behaviors that `prefetch --hydrate` has today
+and also adds the specified files and folders to the `always-hydrate` file, so that those files will
+continue to be hydrated in the future.
```

Comment:
Possible UX feature:

`--always-hydrate` could use `git ls-files` to check how many files (approximately) will by hydrated (before staring any of the prefetch work).

We could also consider warning the user if the number is too large (to prevent them from unintentionally hydrating their entire repo).

Reply:
What is too large? We actually currently allow a user to prefetch and hydrate all with `--files * --hydrate`, and I don't see a strong argument against `--files * --always-hydrate`. In fact, some of the folks in Windows and Office who routinely build the entire repo will probably want exactly that behavior.

The one protection that we currently have in place, and was technically a breaking change when it went in because people kept doing it on accident, is that running `gvfs prefetch` would download all blobs and `gvfs prefetch --hydrate` would hydrate all files. Now they fail (not warn!) and force you to specify `--files *` to indicate that you really mean it. We'll definitely extend that same protection to `--always-hydrate`. Beyond that, it's harder for me to see what would be a reasonable limit.

Reply:
That makes sense, thanks!